Amend explicitAttrs Processing

jason-fox opened the pull request:
~Introduces optional hidden attributes to device provisioning. If an attribute is marked as `hidden`, it may be used for expression parsing, but is **not** passed into the context broker.~ Moves explictAttrs processing closer to the northbound CB request

This fixes #1151 , which can't **currently** be done using `explicitAttrs` since undefined attrs cannot be used in expressions.

- Added code to NGSI-v2 processing
- Added code to NGSI-LD processing
- Added tests for NGSI-v2 and NGSI-LD
- Updated documentation.

Whilst creating this PR I noticed JEXL processing was not tested using NGSI-LD.( I needed a test for `hidden` attrs) so I added in the missing tests and in doing so I also discovered a small processing bug. #1118 wasn't extended to JEXL calculated attributes, so it was still possible to attempt to send `null` to an NGSI-LD broker.

~The reason this change cannot be placed in a plug-in, is that  the removal of `hidden` attributes needs to occur on **Entity creation** as well as update, hence it can't be  just a simple updatePlugin. Also it can't be `explicitAttrs` with unnamed attributes in the measure, since those are removed before the expression plugin kicks in.~

jason-fox commented:
#### Sample provisioning.

Provision a service group in the usual manner:

```console
curl -L -X POST 'http://localhost:4041/iot/services' \
-H 'fiware-service: openiot' \
-H 'fiware-servicepath: /' \
-H 'Content-Type: application/json' \
--data-raw '{
 "services": [
   {
     "apikey":      "4jggokgpepnvsb2uv4s40d59ov",
     "cbroker":     "http://orion:1026",
     "entity_type": "Motion",
     "resource":    "/iot/json"
   }
 ]
}'
```

Provision a device with several `hidden` attributes:

```console
curl -L -X POST 'http://localhost:4041/iot/devices' \
-H 'fiware-service: openiot' \
-H 'fiware-servicepath: /' \
-H 'Content-Type: application/json' \
--data-raw '{
    "devices": [
        {
            "device_id": "52",
            "protocol": "GENERIC_PROTO",
            "entity_name": "WasteContainer:WC52",
            "entity_type": "WasteContainer",
            "expressionLanguage": "jexl",
            "explicitAttrs":true,
            "attributes": [
                { "name": "TimeInstant", "type": "DateTime", "expression": "ts|toisodate" },
                {
                    "name": "location",
                    "type": "geo:json",
                    "expression": "{coordinates: [longitude,latitude], type: '\''Point'\''}"
                },
                {
                    "name": "fillingLevel",
                    "type": "Number",
                    "expression": "level / 100"
                },
                {
                    "name": "level",
                    "type": "Number"
                },
                {
                    "name": "latitude",
                    "type": "Number",
                    "hidden": true
                },
                {
                    "name": "longitude",
                    "type": "Number",
                    "hidden": true
                },
                {
                    "name": "ts",
                    "type": "Number",
                    "hidden": true
                }
            ]
        }
    ]
}'
```

#### Result - only attributes without `hidden` are sent to the context broker

At this point all attributes are `null` except for `TimeInstant` which is a `null` cast to an ISODate.
As expected `TimeInstant` is propagated as metadata in the usual manner. `longitude`, `latitude` and `ts` are not sent.

```json
{
    "id": "WasteContainer:WC52",
    "type": "WasteContainer",
    "TimeInstant": {
        "type": "DateTime",
        "value": "1970-01-01T00:00:00.000Z",
        "metadata": {}
    },
    "fillingLevel": {
        "type": "Number",
        "value": null,
        "metadata": {}
    },
    "level": {
        "type": "Number",
        "value": null,
        "metadata": {}
    },
    "location": {
        "type": "geo:json",
        "value": null,
        "metadata": {}
    }
}
```

Send a measure

```console
curl -L -X POST 'http://localhost:7896/iot/json?k=4jggokgpepnvsb2uv4s40d59ov&i=52' \
-H 'Content-Type: application/json' \
--data-raw '{"latitude": 3, "longitude": 20, "level": 30, "ts" : 1}'
```

#### Result - calculated values are displayed:

`longitude` , `latitude` and `ts` are used in the calculations, but are not sent to the context broker.

```json
{
    "id": "WasteContainer:WC52",
    "type": "WasteContainer",
    "TimeInstant": {
        "type": "DateTime",
        "value": "1970-01-01T00:00:00.001Z",
        "metadata": {}
    },
    "fillingLevel": {
        "type": "Number",
        "value": 0.3,
        "metadata": {
            "TimeInstant": {
                "type": "DateTime",
                "value": "1970-01-01T00:00:00.001Z"
            }
        }
    },
    "level": {
        "type": "Number",
        "value": 30,
        "metadata": {
            "TimeInstant": {
                "type": "DateTime",
                "value": "1970-01-01T00:00:00.001Z"
            }
        }
    },
    "location": {
        "type": "geo:json",
        "value": {
            "coordinates": [
                20,
                3
            ],
            "type": "Point"
        },
        "metadata": {
            "TimeInstant": {
                "type": "DateTime",
                "value": "1970-01-01T00:00:00.001Z"
            }
        }
    }
}
```

mapedraza commented:
Hi Jason, thank you for contributing this new functionality. At this point, instead of adding more complexity to the API with another new parameter, Why don't just fixing `explicitAttrs` functionality from the beginning?

In particular, instead of appliying the explicitAttr filtering at southbound (i.e. to the measures being recieved by IOTA), move the filtering to the northbound (i.e. at the point in which the CB update request is built). That way every measure/attribute will be available in calculatings done by JELX expressions, but only the ones explicetly provisioned will pass to the CB

jason-fox commented:
@mapedraza - fixed and refactored. It looks like code for stripping extra attrs was only present in the Ultralight and JSON IoT Agents. I've removed that in two other PRs  and pointed to my branch for now and all tests are still green.

`'TimeInstant` was treated differently to everything else in the deleted code, its exceptional case has been copied to this PR in that `'TimeInstant` is left as an attribute even when not explicit. This allows  measures to provide their own `TimeInstant`.

I think this PR is now ready for review. I needs to process before telefonicaid/iotagent-json#605 and telefonicaid/iotagent-ul#523 of course.

mapedraza commented on lib/services/ngsi/entities-NGSI-v2.js:
```diff
@@ -323,6 +324,7 @@ function sendQueryValueNgsi2(entityName, attributes, typeInformation, token, cal
  */
 function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, callback) {
     let payload = {};
+    const explicitAttrsList = ['TimeInstant'];
```
Maybe is a good idea to move this part where all the timeinstant is managed (even if this part of the code needs to be refactored in the future): 

Here:
https://github.com/telefonicaid/iotagent-node-lib/blob/ddbcd33094e64003da4cacfd0c2743cd723e437f/lib/services/ngsi/entities-NGSI-v2.js#L392-L406

And here:
https://github.com/telefonicaid/iotagent-node-lib/blob/ddbcd33094e64003da4cacfd0c2743cd723e437f/lib/services/ngsi/entities-NGSI-v2.js#L428-L440

Further, this is declared as a "magic string". It is not a good pattern to follow. Using the constant defined for this purpose is a better idea:

https://github.com/telefonicaid/iotagent-node-lib/blob/6748fc9cedee6133d530e7fd37581236e8d369ec/lib/constants.js#L56

jason-fox replied:
Magic number replaced by constant - Fixed b707d37

mapedraza commented on test/unit/ngsiv2/examples/contextRequests/updateContextHiddenAttrsPlugin.json:
```diff
@@ -0,0 +1,16 @@
+{
+  "location": {
+    "type": "geo:json",
+    "value": {
+      "coordinates": [
+        9.704217,
+        50.5744
+      ],
+      "type": "Point"
+    }
+  },
+  "TimeInstant": {
+    "type": "DateTime",
+    "value": "2021-11-18T14:20:14.901Z"
+  }
+}
```
I am not able to see where is used this file, can you confirm me it is actually used?

jason-fox replied:
Fixed - deleted 7fecd2e

mapedraza commented on lib/services/ngsi/entities-NGSI-v2.js:
```diff
@@ -323,6 +324,7 @@ function sendQueryValueNgsi2(entityName, attributes, typeInformation, token, cal
  */
 function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, callback) {
     let payload = {};
+    const explicitAttrsList = ['TimeInstant'];
```
Given the change on TimeStamp behavior, a test that cover and checks JEXL expressions + explicitAttrs + autogenerated Timestamp and provisioned Timestamp should be included

jason-fox replied:
Fixed 39c4744 - NGSI-v2 and NGSI-LD tests added.

mapedraza replied:
Correct me if I am wrong, with the changes included in this PR, if you send a measure, `TimeInstant:foo`, using the flag `timestamp:false`, this measure would progress to the entity because of declaring `const explicitAttrsList = ['TimeInstant'];`

jason-fox replied:
The [relevant code](https://github.com/telefonicaid/iotagent-ul/blob/master/lib/commonBindings.js#L127) in the original can be found in `manageConfigurationRequest(apiKey, deviceId, device, objMessage)` - which is due to be deleted. If you supply `TimeInstant' it is propagated through to the context broker.

Incidentally this is where the original _"Magic Number"_ was taken from.

jason-fox replied:
As in the example above, you can generate a `TimeInstant` from an expression and supply a unix timestamp rather than an ISO date too:

```json
"attributes": [
                { "name": "TimeInstant", "type": "DateTime", "expression": "ts|toisodate" }
]
```

```console
curl -L -X POST 'http://localhost:7896/iot/json?k=<apikey>&i=<deviceId>' \
-H 'Content-Type: application/json' \
--data-raw '{"latitude": 3, "longitude": 20, "level": 30, "ts" : 1}'
```

mapedraza replied:
What I am highlighting here is different.

If you provision an entity like this (note the option [`"timestamp":false`](https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/api.md#device-model)):

```JSON
...
 "timestamp":false
 "active": [
     { "name": "attr",  "type": "number" }
 ],
 "explicitAttrs": true
...
```

And you send a measure like this one:

```JSON
 {"TimeInstant": "foo"}
```

The resulting resulting update operation would include an attribute `TimeInstant` even if it was not defined in the provision and `explicitAttrs` value was `true`

jason-fox replied:
Fixed 710fabe - test case added.

jason-fox commented on lib/services/ngsi/entities-NGSI-v2.js:
```diff
@@ -402,6 +416,13 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
+                            hidden.forEach((attr) => {
+                                delete result[attr];
+                            });
+                        }
```
Fixed - multi-entity tests added 79523fa

jason-fox commented on lib/services/ngsi/entities-NGSI-v2.js:
```diff
@@ -402,6 +416,13 @@ function sendUpdateValueNgsi2(entityName, attributes, typeInformation, token, ca
+                            const hidden = _.difference(_.keys(result), explicitAttrsList);
+                            hidden.forEach((attr) => {
+                                delete result[attr];
+                            });
```
That worked fine for `hidden`, but the logic in`explicitAttrs` is the other way up. What I need to remove is elements in set B which are not in set A, so first I have to define "set A " using `typeInformation` (in one loop) and then only run the delete on the diff. (the keys of "set B" minus the keys of "set A").  The logic is now nicely encapsulated into a function so you could amend it yourself if you feel it really necessary.

mapedraza replied:
After discussing with the rest of the team in other channels, I raised a new issue, #1168 for discussing the idea of removing underscore from codebase in the future. This will not block this PR.

NTC

mapedraza commented on test/unit/ngsiv2/examples/contextRequests/updateContextMultientityPlugin15.json:
```diff
@@ -0,0 +1,25 @@
+{
+  "actionType": "update",
+  "entities": [
+    {
+      "id": "gps1",
+      "type": "GPS"
+    },
+    {
+      "explicit": {
+        "type": "number",
+        "value": 52
+      },
+      "type": "GPS",
+      "id": "SO5"
+    },
+    {
+      "explicit": {
+        "type": "number",
+        "value": 13
+      },
+      "type": "GPS",
+      "id": "SO6"
+    }
+  ]
+}
```
This is a particular case in which the main entity does not have any attribute. I raised an Issue in order to discuss if this makes sense or not (https://github.com/telefonicaid/iotagent-node-lib/issues/1167)

NTC

mapedraza commented on lib/services/ngsi/entities-NGSI-v2.js:
```diff
@@ -312,6 +313,72 @@ function sendQueryValueNgsi2(entityName, attributes, typeInformation, token, cal
+function removeHiddenAttrs(result, typeInformation) {
+    delete result.id;
+    delete result.type;
+    console.error();
```
Debug leftover?

jason-fox replied:
Yes. Fixed fe2b829

mrutid replied:
LGTM

mrutid reviewed (approved):
LGTM

mapedraza reviewed (approved):
LGTM

mapedraza commented:
The merging strategy to follow is:
1. Edit the PRs on the agents first (https://github.com/telefonicaid/iotagent-ul/pull/523 and https://github.com/telefonicaid/iotagent-json/pull/605) pointing again to telefonicaid repo. After this PRs CI tests will fail. 
2. Merge this PR (iotagent-node-lib)
3. Rerun CI tests on agents and merge after they pass all the checks

jason-fox commented:
> The merging strategy to follow is:
> 
> 1. Edit the PRs on the agents first ([Remove preprocess stripping of explicitAttrs iotagent-ul#523](https://github.com/telefonicaid/iotagent-ul/pull/523) and [Remove preprocess stripping of explicitAttrs iotagent-json#605](https://github.com/telefonicaid/iotagent-json/pull/605)) pointing again to telefonicaid repo. After this PRs CI tests will fail.
> 2. Merge this PR (iotagent-node-lib)
> 3. Rerun CI tests on agents and merge after they pass all the checks

Reversions completed. Tests are now _"failing successfully"_ :smile:

fgalan commented on lib/services/ngsi/entities-NGSI-LD.js:
```diff
@@ -403,6 +403,70 @@ function addLinkedEntities(typeInformation, json) {
     });
```
CHANGES_NEXT_RELEASE entry is missing in this PR.

jason-fox replied:
Fixed 8f7e8fa - The PR was split in two and a second CNR was not added.
